# BetaReconstruct
BetaReconstruct: A generative method for ancestral sequence reconstruction

## Abstract:
Ancestral sequence reconstruction (ASR) is a foundational task in evolutionary biology, offering insights into the molecular past and guiding studies of protein function and adaptation. Traditional ASR methods rely on a multiple sequence alignment (MSA), a phylogenetic tree, and an evolutionary model. However, the assumed underlying alignments and trees are often uncertain and the model often focuses on substitutions, while the insertion and deletion (indel) evolutionary process is not explicitly accounted for. Here, we introduce BetaReconstruct, a novel generative approach to ASR that leverages recent advances in natural language processing (NLP) and hybrid transformer architectures. Our method is trained on large-scale simulated datasets with gold-label ancestral sequences and later applied to real-world protein sequences without requiring MSAs or phylogenetic trees. We demonstrate that BetaReconstruct generalizes effectively to diverse evolutionary scenarios and accurately reconstructs ancestral sequences from empirical mammalian data. This work presents a scalable and alignment-free strategy for ASR, underscoring the power of data-driven models to capture complex evolutionary signals beyond the reach of traditional methods.

![alt text](https://github.com/technion-cs-nlp/BetaReconstruct/blob/main/outline_image.png)
Illustration of ASR prediction using BetaReconstruct. (a): Consider the evolution of the sequence “AAMM”, to the proteins: “AAM”, “AYM” and “ATMMM”. (b) BetaReconstruct pipeline. (Ⅰ): The input to the model, unaligned protein sequences; (Ⅱ): The protein sequences are concatenated with special characters between; (Ⅲ): The model processes the input and predicts the output, which is the ancestral sequence (Ⅳ).
